### PR TITLE
refactor(maven): Prepare for hybrid fetching approach

### PR DESCRIPTION
**Pull request by zharinov**
## Changes:

- Introduce intermediate `ReleaseMap` structure
- Make current approach (based on HEAD requests) fill the release map values of type `Release`
- Return non-null values of `ReleaseMap` values

## Context:

Ref #12576

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/maven/index.ts [hunk 1]
```diff
@@ -145,126 +124,139 @@ async function getSnapshotFullVersion(
...
 
-  logger.debug(`Looking up ${dependency.display} in repository ${repoUrl}`);
-
-  let releases: Release[] = null;
-  const versions = await getVersionsFromMetadata(dependency, repoUrl);
-  releases ||= await getReleasesFromHeadRequests(dependency, repoUrl, versions);
-  releases ||= getFallbackReleases(versions);
+  logger.debug(`Looking up ${mvnDep.display} in repository ${repoUrl}`);
 
+  let releaseMap = await fetchReleasesFromMetadata(mvnDep, repoUrl);
```
Comment (zharinov) [inline lib/datasource/maven/index.ts:245]: Here we'll add HTML-based release info fetching, which will handle most of the Maven Central versions

**Comment by rarkins** [comment]
Variable renaming makes this harder to review and should be removed or separated

File: lib/datasource/maven/index.ts [hunk 2]
```diff
@@ -166,74 +156,83 @@ async function createUrlForDependencyPom(
...
+  );
 
-    artifactsInfo = results.reduce(
-      (acc, [key, value]) => ({
-        ...acc,
-        [key]: value,
-      }),
-      {}
-    );
+  if (!workingReleaseMap) {
```
Comment (zharinov) [inline lib/datasource/maven/index.ts:177]: The whole branch does only cache population

File: lib/datasource/maven/index.ts [hunk 3]
```diff
@@ -166,74 +156,83 @@ async function createUrlForDependencyPom(
...
+          const artifactUrl = getMavenUrl(dependency, repoUrl, pomUrl);
+          const res = await isHttpResourceExists(artifactUrl);
+          const release: Release = { version };
+
+          if (is.string(res)) {
+            release.releaseTimestamp = res;
+          }
+
+          // Retry earlier for error status other than 404
+          if (res === null) {
```
Comment (zharinov) [inline lib/datasource/maven/index.ts:209]: `null` stands for any error other than 404 or `ENOTFOUND`. Probably, it's time to refactor `isHttpResourceExists` function to return `type ExistenceCheckResult = 'exists' | 'not-found' | 'error'` instead. What do you think?

File: lib/datasource/maven/index.ts [hunk 4]
```diff
@@ -166,74 +156,83 @@ async function createUrlForDependencyPom(
...
-  const result = versions?.map((version) => ({
-    version,
-  }));
-  return result ?? null;
+function getReleasesFromMap(releaseMap: ReleaseMap): Release[] {
+  const releases = Object.values(releaseMap).filter(Boolean);
+  if (releases.length) {
+    return releases;
+  }
+  return Object.keys(releaseMap).map((version) => ({ version }));
```
Comment (zharinov) [inline lib/datasource/maven/index.ts:235]: We either return some results from HEAD responses (with further PR, to be combined with HTML-based results) or just generate from `maven-metadata.xml` without any timestamp.

The end result will be either ALL releases are having timestamp or NONE of them.

**Comment by zharinov** [comment]
(ready for review)

File: lib/datasource/maven/index.ts [hunk 5]
```diff
@@ -166,74 +156,83 @@ async function createUrlForDependencyPom(
...
+          // Retry earlier for error status other than 404
+          if (res === null) {
+            retryEarlier = true;
+          }
+
+          workingReleaseMap[version] = res ? release : null;
+        }
+      );
+
+      await pAll(queue, { concurrency: 5 });
```
Comment (viceice) [inline lib/datasource/maven/index.ts:217]: not sure if this will cause too many concurrent connections?

Comment (zharinov) [reply lib/datasource/maven/index.ts:217]: It meant to be only 5 at the particular moment of time. My intention was to preserve current behavior, but rewrite with imperative `pAll` that looks more readable (at least, to me personally).

**Review by viceice — APPROVED** [review]

**Pull request merged**